### DISCUSSION TITLE
Dynamically import components depending on getStaticProps response

### DISCUSSION THREAD

**Discussion by urko-pineda** [vercel/next.js#11291] [Help] [answered]
We have a CMS where we get all the components that a page has. The components are on the project's `/components/*` directory. Here is an example:

```js
import React from 'react'
import dynamic from 'next/dynamic'
import fetch from 'isomorphic-fetch'

const Page = (props) => {
  const {
    id,
    data,
  } = props

  const components = data.map((d) => dynamic(() => import(`../../components/${d}`)))

  return (
    <>
      <p>
        Page {id} has the next components: {data.map((d) => `"${d}", `)}
      </p>
      {components.map((Component, i) => (
        <div key={i}>
          <Component />
        </div>
      ))}
    </>
  )
}

export const getStaticProps = async ({ params }) => {
  const {
    id,
  } = params
  const result = await fetch(`http://localhost:4000/get?id=${id}`)
  // Here is an array with the components.
  const data = await result.json()
  return ({
    props: {
      id,
      data,
    }
  })
}

export const getStaticPaths = async () => {
  const result = await fetch('http://localhost:4000/get/all')
  const data = await result.json()
  const paths = data.map((id) => `/p/${id}`)
  return ({
    paths,
    fallback: false,
  })
}

export default Page
```

The problem is that when this page is exported, the components are not embedded into the exported HTML page. This is a normal behaviour because we import dynamically the components on render, because we depend on the data from `getStaticProps`.

**The question is, there is away to make Next.js aware of what components we need in the `getStaticProps`? This way, when the page is exported based on `getStaticProps`, we could tell Next.js to include in the exported HTML the components needed on that page.**

**IMPORTANT NOTE:** We would like to avoid importing all the components of the project in the page. This is a solution but we have a lot of components and each page does not use all of them, so, the exported HTML would be too large.

Thanks!

**Comment by dotlouis** [comment]
Hello,
I have the same use-case
Right now, we lose all benefits of Server Side Generation because of this.
I even tried serializing the component's code and passing it as a static prop (which I know is weird and ugly).
Obviously, it did not work because the code was processed by webpack and evaluated in a completely different context. Although this is the only idea I could come up with.
Any takers?

**Comment by timneutkens** [comment, marked as answer]
Using ```import(`../../components/${d}`)``` causes **every single file under that directory** to be bundled as a separate file. In 99% of the cases this is not what you want as it causes significant performance issues with regards to builds.

Instead of doing 

```jsx
const components = data.map((d) => dynamic(() => import(`../../components/${d}`)))
```

You'll want to create a list of components at the top level of the module:

```jsx
const componentList = {
  MyComponent: dynamic(() => import(`../../components/my-component`)),
  OtherComponent: dynamic(() => import(`../../components/other-component`))
}
```

And then use it:

```jsx
const components = data.map((d) => componentList[d]))
```

This would look something like:

```jsx
import dynamic from 'next/dynamic'

const componentList = {
  MyComponent: dynamic(() => import(`../../components/my-component`)),
  OtherComponent: dynamic(() => import(`../../components/other-component`))
}

export default function HomePage({data}) {
  const components = data.map((d) => componentList[d]))
  return <>
    {components.map(Component => {
      return <li>
       <Component />
      </li>
    })}
  </>
}
```

Though it seems counter-intuitive this gives the best result for what you're trying to do. Creating the list of components causes you to only import the components that are really needed, instead of the whole `components` directory.

**Reply by vimota** [reply]
I'm having the same issue - I'm trying to move from having my markdown blog pages rendered on the client-side to having the statically generated. I see what you're suggesting here seems to make sense for components, but for my case of posts I think it should be fine, no? 

Is there any way to achieve this without requiring to list every post?

```
export const getStaticProps: GetStaticProps = async context => {
  const { postSlug } = context.params;
  const postMetadata = posts.find(x => x.slug == postSlug);
  const MdComponent = dynamic(() =>
    import(`../../content/posts/${postSlug}.md`)
  );
  return { props: { postSlug, content: renderToString(<MdComponent/>), postMetadata } };
};
```

![image](https://user-images.githubusercontent.com/865701/79298344-bdeeeb80-7e95-11ea-9c0d-381f822dbef5.png)

**Reply by vimota** [reply]
Actually I was able to resolve this by doing:

```
export const getStaticProps: GetStaticProps = async context => {
  const { postSlug } = context.params;
  const postMetadata = posts.find(x => x.slug == postSlug);
  const filePath = path.join(process.cwd(), `./content/posts/${postSlug}.md`);
  const md = fs.readFileSync(filePath, 'utf8');
  const html = await markdownToHtml(md);
  return { props: { postSlug, content: html, postMetadata } };
};
```

And using the markdowntohtml you guys setup for the Next site 🙌 https://github.com/zeit/next-site/blob/master/lib/docs/markdown-to-html.js#L46

**Reply by timneutkens** [reply]
Glad you found the code to do it 👍 This was going to be my recommendation indeed.

It's covered by this example too: https://github.com/zeit/next.js/blob/canary/examples/blog-starter/pages/posts/%5Bslug%5D.js#L49-L69

**Reply by thomaszdxsn** [reply]
When use .mdx for blog post, is the dynamic import are only choice?

**Reply by odensc** [reply]
Having the same question re. importing MDX files using `dynamic`. I know you're not supposed to use template strings with dynamic imports, but it's the only way I could find to get this to work (and only partially - SSR still doesn't work). 

What I'm doing is something like this:

```tsx
const Article = ({ filename }: ArticleProps) => {
	const MDX = dynamic(() =>
		import(`../../content/articles/${filename}`)
	);
	
	return (
		<Content>
			<ArticleBody>
				<MDXProvider components={mdxComponents}>
					<MDX />
				</MDXProvider>
			</ArticleBody>
		</Content>
	);
};
```

Is there a better way to do this and preserve SSR/static builds?

**Reply by sreetamdas** [reply]
Same as @odensc mentioned; using dynamic imports causes a visual break in the time that it takes for the `mdx` component to be imported; instead, is there a way to pass the mdx component as a `prop` itself like we do for plain `.md` files?

At the moment, NextJS doesn't allow non-JSON serializable content to be passed as props iirc.

**Reply by odensc** [reply]
@sreetamdas - I ended up using [next-mdx-enhanced](https://github.com/hashicorp/next-mdx-enhanced) as mentioned in another issue I found. It's really quite overkill for me, and I find the `getStaticProps` way of doing things to be much cleaner, so it's not ideal - but it works at least.

**Reply by germanliu** [reply]
I have the same issue. Is there a way for this to work with server side rendering?

**Reply by FBosler** [reply]
@timneutkens Thx for the answer. I arrived at the same point and converted my `dynamic` imports based on a user-input into a pre-defined map of the components. Just as in your example. And this works wonderfully for cases where you have a manageable amount of components. 

But what when you have hundreds or thousands of components? Does it make sense and is it possible to generate that component map at build time dynamically?

Something like:
```javascript
const Home = ({ componentList }) => {
    const [compName, setCompName] = useState("");

    return (
        <div>
            <form
                onSubmit={(e) => {
                    e.preventDefault();
                    console.log(compName);
                }}
            >
                <input type="text" onChange={(e) => setCompName(e.target.value)}></input>
            </form>
            {componentList[compName]}
        </div>
    );
};

export const getStaticProps = async () => {
    const result = await fetch(DATABASE); // fetch list of all components
    // Here is an array with the components.
    const data = await result.json();

    const componentList = {};

    data.forEach((entry) => {
        componentList[entry.name] = dynamic(() => import(`${entry.path}`));
    });

    return {
        props: componentList,
    };
};

export default Home;
```

When trying this, I am getting

> SerializableError: Error serializing `.bye.$$typeof` returned from `getStaticProps` in "/".

So it would appear this is not possible, is there another approach to achieve the desired result without writing a component list with hundreds of entries?

**Reply by cjimmy** [reply]
@timneutkens Thank you for the answer! This solved an issue for me where it appeared CSS was being mixed up in production (but not development). Considering how powerful dynamic importing is, I think many developers come to the same conclusion the OP came to – that you can simply import a component based on a var set elsewhere. It works well enough for a while, but ultimately is bad practice. I wonder if it's worth calling out in the docs on dynamic importing, that dynamically importing using a var results in every file generated separately (<- I think I'm butchering that because honestly, I'm not exactly sure what's happening underneath).

**Reply by danielhv10** [reply]
It worked well, thanks for the solution 😄

**Reply by iankduffy** [reply]
For this, I see a flash with CSS (CSS modules) which is causing a CLS issue, is there any way to stop this?

**Reply by mohitk05** [reply]
Hello @timneutkens, thanks for the solution. I have a similar use-case as that posted by Urko where I need to map data to components. As you mentioned, doing it using template string will lead to separate chunks being generated for each component.
I tried implementing your solution with a list of components with dynamic imports, but I can still see that each component has a separate chunk. Is there a way to have a chunk per page such that it contains JS of all components being used in a particular page? This seems to be the best solution in terms of performance.

**Comment by germanliu** [comment]
I have a similar use case. The component I want to dynamically import depends on a database field "type" that I get in getStaticProps. However, I can't find a way to use that "type" field to conditionally render a component using dynamic import AND preserve server side rendering.

Any ideas on how to approach this are welcome!

**Reply by Upguard** [reply]
Did you ever find a way to solve this?

**Comment by eric-burel** [comment]
Bumping in the same issue, I can't find a valid answer for MDX. I don't think the accepted answer actually solves this, because the component map itself is not dependent on getStaticProps response, it is predefined, so you have to write the list of possible pages by hand.
I can't compute it on the fly as I don't seem to have access either to async calls or to `fs` itself at the page top_level.

~~Next-mdx-enhanced is quite limited because the config happens at `next.config` level instead of page level. It's not very scalable when you have different type of MDX content on your website. I had kind of a similar issue with react-static markdown plugins and dropped it for this reason~~ (edit: actually it seems way more powerful than I first understood).

Best solution I can think of is... just using MDX in this scenario, so creating the page in MDX directly. And for regular MD files use the accepted solution to get them as string.


**Edit: to sum it up, for the generic issue asked in the initial question:**

- String conversion to HTML is limited, as you lack customization possibilities. MDX has the advantage that it create a React component using context, and you can customize the layout quite easily.
- Any solution with `dynamic` will be... dynamic. So no server side render (as far as I understand). They also require the components to be known in advance, so no possibility to load a whole directory for instance.
- next-mdx-enhanced seems to be an interesting alternative. To be tested.

I think that the underlying issue this questions and further comments raise, is that you can't really build components of the page on the fly, only the page itself.
That would probably mean introducing a concept of "page fragments" as well as whole page, or a kind of `dynamic` but for the static render.

For the record, for my small use case, I've simply created an `entries.ts` file where I import all my `.md(x)` to make them React component. Then I load the right component depending on the pageName. Adding `dynamic` would give a bit more scalability for an higher number of pages, price would be no server render of the page (which is not optimal for static markdown content).

**Edit 2 to sum it up, for MDX specific use case**
For MDX usage I confirm that [`next-mdx-enhanced`](https://github.com/hashicorp/next-mdx-enhanced) is **AWESOME**. 
- You can get the list of doc files to create an index using `getStaticProps` + `fs` as usual
- If you need the frontmatter for this index, you can also use the more advanced approach recommended in `next-mdx-enhanced` readme, with [babel-plugin-import-glob-array](https://github.com/jescalan/babel-plugin-import-glob-array). That could be slower for a very huge number of files though?
- You can use the `layout` frontmatter property to define your page layout, and even customize for specific page using advanced features.
- You can guess the right layout on the fly based on the `__resourcePath` (path of the .md file), in the `extendFrontMatter` function => you don't even need to set the layout for  each page

So it's a great fit for me, including in the scenario where you have various types of MD content. See our [next.config.js](https://github.com/VulcanJS/vulcan-next-starter/blob/devel/next.config.js#L40), and the [doc index page](https://github.com/VulcanJS/vulcan-next-starter/blob/devel/src/pages/docs/index.tsx).

**Comment by danoc** [comment]
Another way to solve the SRR dynamic import MDX issue is to `import` the `*.mdx` file within `getStaticProps`/`getSeverSideProps` and then use `ReactDOMServer.renderToStaticMarkup(<MyImportedMDXComponent />)` to get the component as a string of HTML. You then send it to the component as a prop called `mdxString`:

```js
const { default: MDXContent, metadata } = await import(
    `../../${foo}/${bar}.mdx`
);

const mdxString = ReactDOMServer.renderToStaticMarkup(
    <MDXContent />
);
```

In the component you use `dynamic` to lazily fetch the MDX. The trick is that you use the `mdxString` as the loading state.

```js
const MDXContent = dynamic(
    () =>
        import(`../../../${foo}/${bar}.mdx`),
    {
        loading: () => <div dangerouslySetInnerHTML={{ __html: mdxString }} />,
    },
);
```

There should be no layout shifting if done correctly.

This approach definitely has some downsides but one benefit is that Fast Refresh will work well as you edit the Markdown file. One issue I ran into with `next-mdx-remote` is that Fast Refresh doesn't work since you'd use `fs` to load local MDX rather than Webpack. (Although maybe using Webpack's `raw-loader` would fix that issue. 🤷)

**Comment by KurtGokhan** [comment]
As I understand, the problem is that Next.js cannot statically analyze which file is being imported inside dynamic, and is not able to make it available in SSR. 

`renderToStaticMarkup` is not ideal because the MDX could have information that is only available in runtime.

What if Next.js would allow passing dynamic modules inside `getStaticProps`. I know the result of `getStaticProps` must be serializable. But is it not technically possible to allow this case?

**Comment by TruRooms** [comment]
Would it be possible using a custom middleware to tackle this problem?

**Comment by TruRooms** [comment]
So here is the solution that I have. Due to the legacy product we have all of our urls in one big `Slug` model in rails that maps urls to models.

For us, that means our `[...slug].js` had to include the root component for each one of those pages eg [product, category, cms] etc. Including them all in the slug page leads to some massive payload sizes. Dynamically importing them was leading to poor lighthouse scores for CLS and LCP.

The solution I worked out this morning has been to use the new middleware feature to dynamically rewrite these slugs. Below is the code:

```
import { NextResponse } from 'next/server'
import cache from 'memory-cache'

// Seachs an api to see if a particulary url slug has a specific type set.
// Then appends the path passed to next with a subfolder so that url can only
// have a specific component to render instead of loading them all in one [...slug].js
// or having to dynamically import them. It also uses a memory cache to not have to make 
// round trips to the api on every request.
export async function middleware(request) {
  const cacheTimeout = 3600 
  const knownBadPaths = ['_next']
  if (knownBadPaths.some(element => request.nextUrl.pathname.includes(element))) {
    let apiRes = cache.get(request.nextUrl.pathname)
    if (apiRes === null) {
      const nextSlug = request.nextUrl.pathname.substring(1).replaceAll('/', ',')
      const apiReq = await fetch(`${process.env.NEXT_PUBLIC_API_ROOT}/public/url_slugs/${nextSlug}`)
      apiRes = await apiReq.json()
      cache.put(request.nextUrl.pathname, apiRes, cacheTimeout)
    }
    if (apiRes.type === 'Category') {
      return NextResponse.rewrite(new URL(`/c${request.nextUrl.pathname}`, request.url))
    } else if (apiRes.type === 'Product') {
      return NextResponse.rewrite(new URL(`/p${request.nextUrl.pathname}`, request.url))
    } //... elseif for any other paths that need to be rewritten for specific types
  }
}

export const config = {
  matcher: '/:path*',
} 
```

This is a first for me. Does anyone see any glaring issues with this implementation?

**Reply by 7iomka** [reply]
@TruRooms Hi. This is your final solution? Is there any changes regardless to new features of middleware now?
